File: mobile/src/utils/tripFare.js
/** Haversine distance in km — same formula as backend */
export function haversineKm(lat1, lng1, lat2, lng2) {
  const R = 6371;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLng = ((lng2 - lng1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLng / 2) *
      Math.sin(dLng / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

/** fare = baseFare + (distanceKm * pricePerKm) */
export function fareFromVehicle(baseFare, pricePerKm, distanceKm) {
  const b = Number(baseFare) || 0;
  const p = Number(pricePerKm) || 0;
  const d = Number(distanceKm) || 0;
  return Math.round((b + d * p) * 100) / 100;
}

File: apps/mobile-flutter/lib/core/utils/trip_fare.dart
import 'dart:math' as math;

double haversineKm(double lat1, double lng1, double lat2, double lng2) {
  const r = 6371.0;
  final dLat = _rad(lat2 - lat1);
  final dLng = _rad(lng2 - lng1);
  final a = math.sin(dLat / 2) * math.sin(dLat / 2) +
      math.cos(_rad(lat1)) * math.cos(_rad(lat2)) * math.sin(dLng / 2) * math.sin(dLng / 2);
  return r * 2 * math.asin(math.sqrt(a));
}

double _rad(double d) => d * math.pi / 180.0;

double fareFromVehicle(Map<String, dynamic> v, double km) {
  final base = (v['baseFare'] as num?)?.toDouble() ?? 10;
  final perKm = (v['pricePerKm'] as num?)?.toDouble() ?? 2;
  return base + perKm * km;
}
